File: src/78-subsets-v2.ts
function subsets(nums: number[]): number[][] {
  let results: number[][] = [[]];
  for (let n of nums) {
    let queue: number[][] = [];

    for(let subset of results) {
      const newSubset = Array.from(subset);
      newSubset.push(n);
      queue.push(newSubset);
    }

    for (let subset of queue) {
      results.push(subset);
    }
  }

  return results;
};

console.log(subsets([1, 2, 3]))

File: src/78-subsets-v2.java
class Solution {
  public List<List<Integer>> subsets(int[] nums) {
    List<List<Integer>> result = new ArrayList<>();
    result.add(new ArrayList<Integer>());

    for (int n: nums) {
      List<List<Integer>> queue = new ArrayList<>();

      for (List<Integer> subset: result) {
        List<Integer> newSubset = new ArrayList<>(subset);
        newSubset.add(n);
        queue.add(newSubset);
      }

      for (List<Integer> subset: queue) {
        result.add(subset);
      }
    }

    return result;
  }
}
